### PR TITLE
refactor: deprecate register api

### PR DESCRIPTION
**Pull request by ncclementi**
Looks like I messed something up in the rebase, I'll fix it soon. 
 
- Closes https://github.com/ibis-project/ibis/issues/8735

### CODE REVIEW

File: ibis/backends/datafusion/tests/test_register.py
Comment (gforsyth) [inline ibis/backends/datafusion/tests/test_register.py:30]:
```suggestion
    with pytest.warns(FutureWarning, match="v9.0"):
        conn.register(tab, "my_table")
    assert conn.table("my_table").x.sum().execute() == 6
```

We should only have the line (or lines) that warns in the context manager block

File: ibis/backends/duckdb/tests/test_client.py
Comment (gforsyth) [inline ibis/backends/duckdb/tests/test_client.py:275]:
```suggestion
        con = ibis.connect(tmp_path / f"out.{extension}")
    t = next(iter(con.tables.values()))
    assert not t.head().execute().empty
```

File: ibis/backends/duckdb/tests/test_register.py
Comment (gforsyth) [inline ibis/backends/duckdb/tests/test_register.py:265]: `warns` should only be around the `con.register` call

File: ibis/backends/tests/test_register.py
Comment (gforsyth) [inline ibis/backends/tests/test_register.py:106]: I think this test should be rewritten to test `read_csv` (and we can probably rename the test file, too)

File: ibis/backends/tests/test_register.py
Comment (gforsyth) [inline ibis/backends/tests/test_register.py:160]: same, this is now a `read_csv` test

File: ibis/backends/tests/test_register.py
Comment (gforsyth) [inline ibis/backends/tests/test_register.py:223]: this should now be a read_parquet test

File: ibis/backends/tests/test_register.py
Comment (gforsyth) [inline ibis/backends/tests/test_register.py:269]: this can be testing reading an iterable of parquet files with `read_parquet`

File: ibis/backends/tests/test_register.py
Comment (gforsyth) [inline ibis/backends/tests/test_register.py:299]: The in-memory tests we can probably leave alone for now, but we should add a comment that these should be updated  to work with the upcoming `read_in_memory` and we should also add a polars dataframe to the mix

File: ibis/backends/tests/test_register.py
Comment (gforsyth) [inline ibis/backends/tests/test_register.py:367]: this can be `read_csv`

**Review by gforsyth — CHANGES_REQUESTED** [review]
Sorry for the delay @ncclementi !

The deprecation implementation is good -- I think some of the tests can be updated now and some should probably wait until we have a standardized `read_in_memory`

File: ibis/backends/tests/test_register.py
Comment (ncclementi) [reply ibis/backends/tests/test_register.py:106]: @gforsyth I noticed that ion the same file we do have already `test_read_csv` and `test_read_csv_glob` similarly for parquet. Do we still want to convert these tests then? what are we testing that is different in this case?

**Comment by ncclementi** [comment]
I modify the tests in question to use `read_csv` or `read_parquet`  I wasn't sure what the name of the test should be or if we need them given that there are separate `read_csv` and `read_parquet`, though the register tests seem to be more complete.  Happy to re-visit.

**Comment by ncclementi** [comment]
**Update regarding adapting `test_register_csv` and similar to use `con.read_csv`**

Turns out that doing this would require to modify the tests quite a bit given that now this will trigger some XPASS failures, and the tests will become way more convoluted than what they are at the moment. 

After chatting with @cpcloud, the suggestion was to leave the tests as they are now, and discuss a decision for when we remove the `register` api completely.

**Comment by gforsyth** [comment]
Ok, I'm good with the changes here, but I think we need a consensus from a few folks on whether to merge now or not.
If we want to add this in Ibis 9.0, I think we also need to handle standardizing `read_in_memory` xref #6593 before we release, because it seems irresponsible to deprecate functionality without offering a replacement.

**Comment by cpcloud** [comment]
@gforsyth Brings up a good point. Let's hold off on merging to 9.0 for now to avoid delaying its release. We can merge this soon after 9.0 is cut (not immediately, just to wait for the bug reports to come in and cut a 9.0.1 or 9.1 bug fix release).

**Comment by ncclementi** [comment]
I just rebased, I think when CI finishes we can get this one in since 9.0 is out

**Comment by ncclementi** [comment]
@gforsyth should we get this one in, or are there still concerns merging this deprecation without having an alternative `read_in_memory` api?

**Comment by ncclementi** [comment]
Ready, waiting for https://github.com/ibis-project/ibis/pull/9251 to be merged.

**Review by cpcloud — APPROVED** [review]

**Pull request merged**
